File: scripts/release.ts
#!/usr/bin/env bun

import { createHash } from "node:crypto";
import { copyFile, mkdir, readFile, writeFile } from "node:fs/promises";
import { basename, join, resolve } from "node:path";
import { createInterface } from "node:readline/promises";

const ROOT = resolve(import.meta.dir, "..");
const CLI_DIR = join(ROOT, "packages", "inferay");
const CLI_PACKAGE_JSON = join(CLI_DIR, "package.json");
const CLI_SOURCE = join(CLI_DIR, "src", "cli.js");
const ELECTROBUN_CONFIG = join(ROOT, "electrobun.config.ts");
const ARTIFACTS_DIR = join(ROOT, "artifacts");
const INSTALLER_DMG = join(ARTIFACTS_DIR, "inferay-installer.dmg");
const PLATFORM_DMG = join(ARTIFACTS_DIR, "inferay-macos-arm64.dmg");
const CHECKSUMS = join(ARTIFACTS_DIR, "checksums.txt");
const NPM_CACHE = "/private/tmp/inferay-npm-cache";
const DEFAULT_RELEASE_REPO = "raymondreaming/inferay";

type Bump = "major" | "minor" | "patch";

interface Options {
	bumpOrVersion: string;
	publishExisting: boolean;
	repo: string;
}

function usage() {
	console.log(`Usage:
  bun run release [patch|minor|major|new|x.y.z] [--repo owner/repo]
  bun run release --resume [--repo owner/repo]

Examples:
  bun run release
  bun run release minor
  bun run release 0.2.0
  bun run release --resume
`);
}

function parseArgs(): Options {
	const args = process.argv.slice(2);
	if (args.includes("--help") || args.includes("-h")) {
		usage();
		process.exit(0);
	}

	const repoIndex = args.indexOf("--repo");
	const repo =
		repoIndex >= 0
			? args[repoIndex + 1]
			: process.env.INFERAY_RELEASE_REPO || DEFAULT_RELEASE_REPO;
	if (!repo) throw new Error("--repo requires owner/repo");

	const positional = args.filter(
		(arg, index) =>
			!arg.startsWith("--") && !(repoIndex >= 0 && index === repoIndex + 1)
	);

	return {
		bumpOrVersion: positional[0] ?? "patch",
		publishExisting:
			args.includes("--resume") || args.includes("--publish-existing"),
		repo,
	};
}

function quoteArg(arg: string) {
	return /[\s"]/g.test(arg) ? JSON.stringify(arg) : arg;
}

async function run(
	cmd: string[],
	options: {
		cwd?: string;
		quiet?: boolean;
		allowFailure?: boolean;
		env?: NodeJS.ProcessEnv;
	} = {}
) {
	if (!options.quiet) console.log(`$ ${cmd.map(quoteArg).join(" ")}`);
	const proc = Bun.spawn(cmd, {
		cwd: options.cwd ?? ROOT,
		stdin: options.quiet ? "ignore" : "inherit",
		stdout: options.quiet ? "pipe" : "inherit",
		stderr: options.quiet ? "pipe" : "inherit",
		env: options.env ?? process.env,
	});
	const exitCode = await proc.exited;
	if (exitCode !== 0 && !options.allowFailure) {
		throw new Error(
			`command failed (${exitCode}): ${cmd.map(quoteArg).join(" ")}`
		);
	}
	return exitCode;
}

async function capture(cmd: string[], options: { cwd?: string } = {}) {
	const proc = Bun.spawn(cmd, {
		cwd: options.cwd ?? ROOT,
		stdout: "pipe",
		stderr: "pipe",
		env: process.env,
	});
	const [stdout, stderr, exitCode] = await Promise.all([
		new Response(proc.stdout).text(),
		new Response(proc.stderr).text(),
		proc.exited,
	]);
	if (exitCode !== 0) {
		throw new Error(
			`command failed (${exitCode}): ${cmd.map(quoteArg).join(" ")}\n${stderr}`
		);
	}
	return stdout.trim();
}

async function promptLine(question: string) {
	const rl = createInterface({
		input: process.stdin,
		output: process.stdout,
	});
	try {
		return (await rl.question(question)).trim();
	} finally {
		rl.close();
	}
}

function parseVersion(version: string) {
	const match = /^(\d+)\.(\d+)\.(\d+)$/.exec(version);
	if (!match) throw new Error(`invalid version: ${version}`);
	return match.slice(1).map(Number) as [number, number, number];
}

function bumpVersion(current: string, requested: string) {
	if (/^\d+\.\d+\.\d+$/.test(requested)) return requested;

	const bump: Bump = requested === "new" ? "patch" : (requested as Bump);
	if (!["major", "minor", "patch"].includes(bump)) {
		throw new Error(
			`expected patch, minor, major, new, or x.y.z; got ${requested}`
		);
	}

	const [major, minor, patch] = parseVersion(current);
	if (bump === "major") return `${major + 1}.0.0`;
	if (bump === "minor") return `${major}.${minor + 1}.0`;
	return `${major}.${minor}.${patch + 1}`;
}

async function readPackageVersion() {
	const pkg = JSON.parse(await readFile(CLI_PACKAGE_JSON, "utf8"));
	if (typeof pkg.version !== "string") {
		throw new Error("packages/inferay/package.json is missing version");
	}
	return pkg.version as string;
}

async function setCliVersion(version: string) {
	const pkg = JSON.parse(await readFile(CLI_PACKAGE_JSON, "utf8"));
	pkg.version = version;
	await writeFile(CLI_PACKAGE_JSON, `${JSON.stringify(pkg, null, "\t")}\n`);

	const source = await readFile(CLI_SOURCE, "utf8");
	const next = source.replace(
		/const VERSION = "\d+\.\d+\.\d+";/,
		`const VERSION = "${version}";`
	);
	if (next !== source) {
		await writeFile(CLI_SOURCE, next);
		return;
	}
	if (!source.includes("version: VERSION")) {
		throw new Error("could not update CLI VERSION constant");
	}
}

async function setDesktopVersion(version: string) {
	const source = await readFile(ELECTROBUN_CONFIG, "utf8");
	const next = source.replace(
		/version:\s*"\d+\.\d+\.\d+"/,
		`version: "${version}"`
	);
	if (next === source) {
		throw new Error("could not update desktop app version");
	}
	await writeFile(ELECTROBUN_CONFIG, next);
}

async function assertCleanGit() {
	const status = await capture(["git", "status", "--short"]);
	if (status) {
		throw new Error(`git worktree must be clean before release:\n${status}`);
	}
}

async function buildArtifacts(version: string) {
	await mkdir(ARTIFACTS_DIR, { recursive: true });
	await run(["bash", "scripts/build-dmg.sh"]);
	await copyFile(INSTALLER_DMG, PLATFORM_DMG);
	await run(
		[
			"npm",
			"pack",
			"--pack-destination",
			"../../artifacts",
			"--cache",
			NPM_CACHE,
		],
		{ cwd: CLI_DIR }
	);
	await writeChecksums([INSTALLER_DMG, PLATFORM_DMG]);
	await run(["hdiutil", "verify", INSTALLER_DMG]);
	await run(["hdiutil", "verify", PLATFORM_DMG]);

	const cliTarball = join(ARTIFACTS_DIR, `inferay-${version}.tgz`);
	console.log(`Created release artifacts:
  ${relative(INSTALLER_DMG)}
  ${relative(PLATFORM_DMG)}
  ${relative(cliTarball)}
  ${relative(CHECKSUMS)}`);
}

async function writeChecksums(paths: string[]) {
	const lines = [];
	for (const path of paths) {
		const hash = createHash("sha256");
		hash.update(await readFile(path));
		lines.push(`${hash.digest("hex")}  artifacts/${basename(path)}`);
	}
	await writeFile(CHECKSUMS, `${lines.join("\n")}\n`);
}

function relative(path: string) {
	return path.startsWith(ROOT) ? path.slice(ROOT.length + 1) : path;
}

async function commitAndTag(version: string) {
	const tag = `v${version}`;
	const existingTag = await run(["git", "rev-parse", "-q", "--verify", tag], {
		quiet: true,
		allowFailure: true,
	});
	if (existingTag === 0) throw new Error(`tag already exists: ${tag}`);

	await run([
		"git",
		"add",
		"electrobun.config.ts",
		"packages/inferay/package.json",
		"packages/inferay/src/cli.js",
	]);
	await run(["git", "commit", "-m", `release ${tag}`]);
	await run(["git", "tag", tag]);
}

async function releaseExists(tag: string, repo: string) {
	const exitCode = await run(["gh", "release", "view", tag, "--repo", repo], {
		quiet: true,
		allowFailure: true,
	});
	return exitCode === 0;
}

async function preflightPublish() {
	await run(["gh", "auth", "status"]);
	let npmUser: string | null = null;
	try {
		npmUser = await capture([
			"npm",
			"whoami",
			"--registry",
			"https://registry.npmjs.org",
			"--cache",
			NPM_CACHE,
		]);
	} catch {
		console.log("npm login required.");
		await run([
			"npm",
			"login",
			"--registry",
			"https://registry.npmjs.org",
			"--cache",
			NPM_CACHE,
		]);
		npmUser = await capture([
			"npm",
			"whoami",
			"--registry",
			"https://registry.npmjs.org",
			"--cache",
			NPM_CACHE,
		]);
	}
	console.log(`npm user: ${npmUser}`);
}

async function publishNpmPackage() {
	const cmd = ["npm", "publish", "--access", "public", "--cache", NPM_CACHE];
	let env = process.env;

	for (let attempt = 1; attempt <= 3; attempt++) {
		const exitCode = await run(cmd, {
			cwd: CLI_DIR,
			allowFailure: true,
			env,
		});
		if (exitCode === 0) return;
		if (attempt === 3) break;

		const answer = await promptLine(
			"Complete npm browser auth and press Enter to retry, or enter an npm OTP: "
		);
		env = answer ? { ...process.env, npm_config_otp: answer } : process.env;
	}

	throw new Error("npm publish failed after auth retries");
}

async function publishRelease(version: string, repo: string) {
	const tag = `v${version}`;
	await preflightPublish();
	await run(["git", "push", "origin", "HEAD", "--tags"]);

	const assets = [PLATFORM_DMG, INSTALLER_DMG, CHECKSUMS];
	if (await releaseExists(tag, repo)) {
		await run([
			"gh",
			"release",
			"upload",
			tag,
			...assets,
			"--repo",
			repo,
			"--clobber",
		]);
	} else {
		await run([
			"gh",
			"release",
			"create",
			tag,
			...assets,
			"--repo",
			repo,
			"--title",
			tag,
			"--notes",
			`Inferay ${tag}`,
		]);
	}

	await publishNpmPackage();
}

async function publishExisting(repo: string) {
	const version = await readPackageVersion();
	const tag = `v${version}`;
	await run(["test", "-f", PLATFORM_DMG]);
	await run(["test", "-f", INSTALLER_DMG]);
	await run(["test", "-f", CHECKSUMS]);
	await publishRelease(version, repo);
	console.log(`Published ${tag}`);
}

async function main() {
	const options = parseArgs();
	if (options.publishExisting) {
		await publishExisting(options.repo);
		return;
	}

	await assertCleanGit();
	const current = await readPackageVersion();
	const next = bumpVersion(current, options.bumpOrVersion);
	const tag = `v${next}`;

	console.log(`Preparing ${tag}`);
	await setCliVersion(next);
	await setDesktopVersion(next);
	await buildArtifacts(next);
	await commitAndTag(next);

	await publishRelease(next, options.repo);
	console.log(`Published ${tag}`);
}

main().catch((error) => {
	console.error(error instanceof Error ? error.message : error);
	process.exit(1);
});

File: native/tooling/src/release.rs
use regex::Regex;
use serde_json::Value;
use sha2::{Digest, Sha256};
use std::env;
use std::fs;
use std::io::{self, Write};
use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};

const BUN_CACHE: &str = "/private/tmp/inferay-bun-cache";
const DEFAULT_RELEASE_REPO: &str = "raymondreaming/inferay";

#[derive(Debug, Clone, PartialEq, Eq)]
struct Options {
    bump_or_version: String,
    publish_existing: bool,
    repo: String,
    help: bool,
}

fn usage() {
    println!(
        "Usage:\n  bun run release [patch|minor|major|new|x.y.z] [--repo owner/repo]\n  bun run release --resume [--repo owner/repo]\n\nExamples:\n  bun run release\n  bun run release minor\n  bun run release 0.2.0\n  bun run release --resume\n"
    );
}

fn parse_args_with_repo_env(args: &[String], repo_env: Option<&str>) -> Result<Options, String> {
    let help = args.iter().any(|arg| arg == "--help" || arg == "-h");
    if help {
        return Ok(Options {
            bump_or_version: "patch".to_owned(),
            publish_existing: false,
            repo: DEFAULT_RELEASE_REPO.to_owned(),
            help: true,
        });
    }
    let repo_index = args.iter().position(|arg| arg == "--repo");
    let repo = match repo_index {
        Some(index) => args.get(index + 1).cloned(),
        None => repo_env
            .map(str::to_owned)
            .or_else(|| Some(DEFAULT_RELEASE_REPO.to_owned())),
    }
    .filter(|repo| !repo.is_empty())
    .ok_or_else(|| "--repo requires owner/repo".to_owned())?;

    let positional = args
        .iter()
        .enumerate()
        .filter(|(index, arg)| {
            !arg.starts_with("--") && repo_index.is_none_or(|repo| *index != repo + 1)
        })
        .map(|(_, arg)| arg.clone())
        .collect::<Vec<_>>();

    Ok(Options {
        bump_or_version: positional
            .first()
            .cloned()
            .unwrap_or_else(|| "patch".to_owned()),
        publish_existing: args
            .iter()
            .any(|arg| arg == "--resume" || arg == "--publish-existing"),
        repo,
        help,
    })
}

fn parse_args(args: &[String]) -> Result<Options, String> {
    parse_args_with_repo_env(args, env::var("INFERAY_RELEASE_REPO").ok().as_deref())
}

fn quote_arg(arg: &str) -> String {
    if arg
        .chars()
        .any(|character| character.is_whitespace() || character == '"')
    {
        serde_json::to_string(arg).unwrap_or_else(|_| arg.to_owned())
    } else {
        arg.to_owned()
    }
}

fn command_text(command: &[String]) -> String {
    command
        .iter()
        .map(|arg| quote_arg(arg))
        .collect::<Vec<_>>()
        .join(" ")
}

fn run_command(
    root: &Path,
    command: &[String],
    cwd: Option<&Path>,
    quiet: bool,
    allow_failure: bool,
) -> Result<i32, String> {
    if !quiet {
        println!("$ {}", command_text(command));
    }
    let (program, arguments) = command
        .split_first()
        .ok_or_else(|| "cannot run an empty command".to_owned())?;
    let mut process = Command::new(program);
    process.args(arguments).current_dir(cwd.unwrap_or(root));
    if quiet {
        process
            .stdin(Stdio::null())
            .stdout(Stdio::null())
            .stderr(Stdio::null());
    } else {
        process
            .stdin(Stdio::inherit())
            .stdout(Stdio::inherit())
            .stderr(Stdio::inherit());
    }
    let status = process
        .status()
        .map_err(|error| format!("failed to run {}: {error}", command_text(command)))?;
    let exit_code = status.code().unwrap_or(1);
    if exit_code != 0 && !allow_failure {
        return Err(format!(
            "command failed ({exit_code}): {}",
            command_text(command)
        ));
    }
    Ok(exit_code)
}

fn capture(root: &Path, command: &[String], cwd: Option<&Path>) -> Result<String, String> {
    let (program, arguments) = command
        .split_first()
        .ok_or_else(|| "cannot run an empty command".to_owned())?;
    let output = Command::new(program)
        .args(arguments)
        .current_dir(cwd.unwrap_or(root))
        .output()
        .map_err(|error| format!("failed to run {}: {error}", command_text(command)))?;
    if !output.status.success() {
        return Err(format!(
            "command failed ({}): {}\n{}",
            output.status.code().unwrap_or(1),
            command_text(command),
            String::from_utf8_lossy(&output.stderr)
        ));
    }
    Ok(String::from_utf8_lossy(&output.stdout).trim().to_owned())
}

fn prompt_line(question: &str) -> Result<String, String> {
    print!("{question}");
    io::stdout().flush().map_err(|error| error.to_string())?;
    let mut answer = String::new();
    io::stdin()
        .read_line(&mut answer)
        .map_err(|error| error.to_string())?;
    Ok(answer.trim().to_owned())
}

fn parse_version(version: &str) -> Result<[u64; 3], String> {
    version
        .split('.')
        .map(|part| {
            part.parse::<u64>()
                .ok()
                .filter(|_| part.bytes().all(|byte| byte.is_ascii_digit()))
        })
        .collect::<Option<Vec<_>>>()
        .and_then(|parts| parts.try_into().ok())
        .ok_or_else(|| format!("invalid version: {version}"))
}

pub(crate) fn bundle_version(version: &str) -> Result<String, String> {
    let invalid = || format!("invalid package version: {version}");
    let [major, minor, patch] = parse_version(version).map_err(|_| invalid())?;
    major
        .checked_mul(1_000_000)
        .and_then(|value| value.checked_add(minor.checked_mul(1_000)?))
        .and_then(|value| value.checked_add(patch))
        .map(|value| value.to_string())
        .ok_or_else(invalid)
}

fn bump_version(current: &str, requested: &str) -> Result<String, String> {
    if parse_version(requested).is_ok() {
        return Ok(requested.to_owned());
    }
    let bump = if requested == "new" {
        "patch"
    } else {
        requested
    };
    if !matches!(bump, "major" | "minor" | "patch") {
        return Err(format!(
            "expected patch, minor, major, new, or x.y.z; got {requested}"
        ));
    }
    let [major, minor, patch] = parse_version(current)?;
    Ok(match bump {
        "major" => format!("{}.0.0", major + 1),
        "minor" => format!("{major}.{}.0", minor + 1),
        _ => format!("{major}.{minor}.{}", patch + 1),
    })
}

pub(crate) fn read_package_version(path: &Path) -> Result<String, String> {
    let package: Value =
        serde_json::from_str(&fs::read_to_string(path).map_err(|error| error.to_string())?)
            .map_err(|error| error.to_string())?;
    package
        .get("version")
        .and_then(Value::as_str)
        .map(str::to_owned)
        .ok_or_else(|| "packages/inferay/package.json is missing version".to_owned())
}

fn set_package_version(package_json: &Path, version: &str) -> Result<(), String> {
    let package_source = fs::read_to_string(package_json).map_err(|error| error.to_string())?;
    let _: Value = serde_json::from_str(&package_source).map_err(|error| error.to_string())?;
    let next_package = replace_json_string_field(&package_source, "version", version)
        .ok_or_else(|| "packages/inferay/package.json is missing version".to_owned())?;
    fs::write(package_json, next_package).map_err(|error| error.to_string())?;

    Ok(())
}

fn replace_capture(source: &str, pattern: &str, value: &str) -> Option<String> {
    let expression = Regex::new(pattern).ok()?;
    let field = expression.captures(source)?.get(1)?;
    let mut next = source.to_owned();
    next.replace_range(field.range(), value);
    Some(next)
}

fn replace_json_string_field(source: &str, field: &str, value: &str) -> Option<String> {
    replace_capture(
        source,
        &format!(r#""{}"\s*:\s*("(?:\\.|[^"\\])*")"#, regex::escape(field)),
        &serde_json::to_string(value).ok()?,
    )
}

fn replace_assignment_version(source: &str, version: &str) -> Option<String> {
    replace_capture(
        source,
        r#"(?m)^\s*version\s*=\s*"([0-9]+\.[0-9]+\.[0-9]+)""#,
        version,
    )
}

fn replace_plist_value(source: &str, key: &str, value: &str) -> Option<String> {
    replace_capture(
        source,
        &format!(
            r"<key>{}</key>\s*<string>([^<]*)</string>",
            regex::escape(key)
        ),
        value,
    )
}

fn set_native_app_version(
    cargo_toml: &Path,
    info_plist: &Path,
    version: &str,
) -> Result<(), String> {
    let cargo = fs::read_to_string(cargo_toml).map_err(|error| error.to_string())?;
    let next_cargo = replace_assignment_version(&cargo, version)
        .ok_or_else(|| "could not update Rust native app version".to_owned())?;
    if next_cargo == cargo {
        return Err("could not update Rust native app version".to_owned());
    }
    let bundle_version = bundle_version(version)?;
    let plist = fs::read_to_string(info_plist).map_err(|error| error.to_string())?;
    let short_version = replace_plist_value(&plist, "CFBundleShortVersionString", version)
        .unwrap_or_else(|| plist.clone());
    let next_plist = replace_plist_value(&short_version, "CFBundleVersion", &bundle_version)
        .unwrap_or(short_version);
    if next_plist == plist {
        return Err("could not update native app Info.plist version".to_owned());
    }
    fs::write(cargo_toml, next_cargo).map_err(|error| error.to_string())?;
    fs::write(info_plist, next_plist).map_err(|error| error.to_string())?;
    Ok(())
}

struct Paths {
    root: PathBuf,
    cli_dir: PathBuf,
    package_json: PathBuf,
    native_app_cargo_toml: PathBuf,
    native_app_info_plist: PathBuf,
    artifacts_dir: PathBuf,
    installer_dmg: PathBuf,
    platform_dmg: PathBuf,
    checksums: PathBuf,
}

impl Paths {
    fn from_root(root: PathBuf) -> Self {
        let cli_dir = root.join("packages/inferay");
        let artifacts_dir = root.join("artifacts");
        Self {
            package_json: cli_dir.join("package.json"),
            native_app_cargo_toml: root.join("native/desktop-host/Cargo.toml"),
            native_app_info_plist: root.join("native/desktop-host/Info.plist"),
            installer_dmg: artifacts_dir.join("inferay-installer.dmg"),
            platform_dmg: artifacts_dir.join("inferay-macos-arm64.dmg"),
            checksums: artifacts_dir.join("checksums.txt"),
            root,
            cli_dir,
            artifacts_dir,
        }
    }

    fn relative(&self, path: &Path) -> String {
        path.strip_prefix(&self.root)
            .unwrap_or(path)
            .to_string_lossy()
            .into_owned()
    }
}

fn command(parts: &[&str]) -> Vec<String> {
    parts.iter().map(|part| (*part).to_owned()).collect()
}

fn assert_clean_git(paths: &Paths) -> Result<(), String> {
    let status = capture(&paths.root, &command(&["git", "status", "--short"]), None)?;
    if status.is_empty() {
        Ok(())
    } else {
        Err(format!(
            "git worktree must be clean before release:\n{status}"
        ))
    }
}

fn write_checksums(paths: &Paths, files: &[&Path]) -> Result<(), String> {
    let mut lines = Vec::new();
    for path in files {
        let bytes = fs::read(path).map_err(|error| error.to_string())?;
        let digest = format!("{:x}", Sha256::digest(bytes));
        let basename = path
            .file_name()
            .ok_or_else(|| format!("invalid artifact path: {}", path.display()))?
            .to_string_lossy();
        lines.push(format!("{digest}  artifacts/{basename}"));
    }
    fs::write(&paths.checksums, format!("{}\n", lines.join("\n")))
        .map_err(|error| error.to_string())
}

fn build_artifacts(paths: &Paths, version: &str) -> Result<(), String> {
    fs::create_dir_all(&paths.artifacts_dir).map_err(|error| error.to_string())?;
    run_command(
        &paths.root,
        &command(&["bash", "scripts/build-dmg.sh"]),
        None,
        false,
        false,
    )?;
    fs::copy(&paths.installer_dmg, &paths.platform_dmg).map_err(|error| error.to_string())?;
    run_command(
        &paths.root,
        &command(&["bun", "pm", "pack", "--destination", "../../artifacts"]),
        Some(&paths.cli_dir),
        false,
        false,
    )?;
    write_checksums(paths, &[&paths.installer_dmg, &paths.platform_dmg])?;
    for dmg in [&paths.installer_dmg, &paths.platform_dmg] {
        run_command(
            &paths.root,
            &[
                "hdiutil".to_owned(),
                "verify".to_owned(),
                dmg.to_string_lossy().into_owned(),
            ],
            None,
            false,
            false,
        )?;
    }
    let tarball = paths.artifacts_dir.join(format!("inferay-{version}.tgz"));
    println!(
        "Created release artifacts:\n  {}\n  {}\n  {}\n  {}",
        paths.relative(&paths.installer_dmg),
        paths.relative(&paths.platform_dmg),
        paths.relative(&tarball),
        paths.relative(&paths.checksums)
    );
    Ok(())
}

fn commit_and_tag(paths: &Paths, version: &str) -> Result<(), String> {
    let tag = format!("v{version}");
    let mut verify = command(&["git", "rev-parse", "-q", "--verify"]);
    verify.push(tag.clone());
    if run_command(&paths.root, &verify, None, true, true)? == 0 {
        return Err(format!("tag already exists: {tag}"));
    }
    run_command(
        &paths.root,
        &command(&[
            "git",
            "add",
            "native/desktop-host/Cargo.toml",
            "Cargo.lock",
            "native/desktop-host/Info.plist",
            "packages/inferay/package.json",
        ]),
        None,
        false,
        false,
    )?;
    run_command(
        &paths.root,
        &[
            "git".to_owned(),
            "commit".to_owned(),
            "-m".to_owned(),
            format!("release {tag}"),
        ],
        None,
        false,
        false,
    )?;
    run_command(
        &paths.root,
        &["git".to_owned(), "tag".to_owned(), tag],
        None,
        false,
        false,
    )?;
    Ok(())
}

fn release_exists(paths: &Paths, tag: &str, repo: &str) -> Result<bool, String> {
    Ok(run_command(
        &paths.root,
        &command(&["gh", "release", "view", tag, "--repo", repo]),
        None,
        true,
        true,
    )? == 0)
}

fn publish_npm_package(paths: &Paths) -> Result<(), String> {
    let base = command(&[
        "bun",
        "publish",
        "--access",
        "public",
        "--cache-dir",
        BUN_CACHE,
    ]);
    let mut otp: Option<String> = None;
    for attempt in 1..=3 {
        let mut invocation = base.clone();
        if let Some(value) = &otp {
            invocation.extend(["--otp".to_owned(), value.clone()]);
        }
        if run_command(&paths.root, &invocation, Some(&paths.cli_dir), false, true)? == 0 {
            return Ok(());
        }
        if attempt < 3 {
            let answer = prompt_line(
                "Complete npm browser auth and press Enter to retry, or enter an npm OTP: ",
            )?;
            otp = (!answer.is_empty()).then_some(answer);
        }
    }
    Err("npm publish failed after auth retries. Authenticate Bun with an npm token via NPM_CONFIG_TOKEN or an .npmrc entry, then run bun run release:resume.".to_owned())
}

fn publish_release(paths: &Paths, version: &str, repo: &str) -> Result<(), String> {
    let tag = format!("v{version}");
    run_command(
        &paths.root,
        &command(&["gh", "auth", "status"]),
        None,
        false,
        false,
    )?;
    run_command(
        &paths.root,
        &command(&["git", "push", "origin", "HEAD", "--tags"]),
        None,
        false,
        false,
    )?;
    let assets = [&paths.platform_dmg, &paths.installer_dmg, &paths.checksums];
    let mut release = if release_exists(paths, &tag, repo)? {
        command(&["gh", "release", "upload", &tag])
    } else {
        command(&["gh", "release", "create", &tag])
    };
    release.extend(
        assets
            .iter()
            .map(|asset| asset.to_string_lossy().into_owned()),
    );
    release.extend(["--repo".to_owned(), repo.to_owned()]);
    if release.get(2).is_some_and(|action| action == "upload") {
        release.push("--clobber".to_owned());
    } else {
        release.extend([
            "--title".to_owned(),
            tag.clone(),
            "--notes".to_owned(),
            format!("Inferay {tag}"),
        ]);
    }
    run_command(&paths.root, &release, None, false, false)?;
    publish_npm_package(paths)
}

fn publish_existing(paths: &Paths, repo: &str) -> Result<(), String> {
    let version = read_package_version(&paths.package_json)?;
    for artifact in [&paths.platform_dmg, &paths.installer_dmg, &paths.checksums] {
        run_command(
            &paths.root,
            &[
                "test".to_owned(),
                "-f".to_owned(),
                artifact.to_string_lossy().into_owned(),
            ],
            None,
            false,
            false,
        )?;
    }
    publish_release(paths, &version, repo)?;
    println!("Published v{version}");
    Ok(())
}

fn repository_root() -> Result<PathBuf, String> {
    let manifest = Path::new(env!("CARGO_MANIFEST_DIR"));
    manifest
        .parent()
        .and_then(Path::parent)
        .map(Path::to_path_buf)
        .ok_or_else(|| "could not resolve repository root".to_owned())
}

pub fn run(args: &[String]) -> Result<(), String> {
    let options = parse_args(args)?;
    if options.help {
        usage();
        return Ok(());
    }
    let paths = Paths::from_root(repository_root()?);
    if options.publish_existing {
        return publish_existing(&paths, &options.repo);
    }

    assert_clean_git(&paths)?;
    let current = read_package_version(&paths.package_json)?;
    let next = bump_version(&current, &options.bump_or_version)?;
    let tag = format!("v{next}");
    println!("Preparing {tag}");
    set_package_version(&paths.package_json, &next)?;
    set_native_app_version(
        &paths.native_app_cargo_toml,
        &paths.native_app_info_plist,
        &next,
    )?;
    build_artifacts(&paths, &next)?;
    commit_and_tag(&paths, &next)?;
    publish_release(&paths, &next, &options.repo)?;
    println!("Published {tag}");
    Ok(())
}
